Add filter_zero_loss_peak argument to spikes_removal_tool and fix one bug.

ericpre opened the pull request:
It should be PR to next_minor, but I initially check out from next_patch... I can change it to next_minor if necessary.
It may also be a good opportunity to add some tests (not for the ui) for the spikes_removal_tool, I guess it should be possible to test the different method/functions without using the ui?

- [x] Add filter_zero_loss_peak argument
- [x] Add documentation
- [x] Add tests
- [x] Fix bug with update pointer position

francisco-dlp commented:
Wouldn't this be equivalent?:


```python

s.isig[start:end].spikes_removal_tool()

```

ericpre commented:
Yes, it is! Thanks :) I will add it to the documentation!

Is there a clever way to set two intervals, that I would have miss? For example to ignore a zero-loss peak?

francisco-dlp commented:
You can pass a ``signal_mask`` to do that. That's simply a boolean matrix. ``Model1D`` has a sort of interactive signal mask editor (e.g. ``set_signal_range`` etc) that would be useful to turn into an independent function to create signal masks more conveniently.

ericpre commented:
Adding a `get_signal_range` method to `Signal1D`, that works in a similar manner than `set_signal_range`  and returns a mask?

francisco-dlp commented:
Something like that. Currently `Model1D` has:

* ``set_signal_range``
* ``remove_signal_range``
* ``reset_signal_range``

They enable editing a mask interactively. Because the mask is stored in the model as an attribute for internal consumption  there is no method to get the mask as you propose.

For the model it makes sense to store it internally, but does it make sense to store it internally for the signal? If not, it could be implemented as a ``Signal`` that includes the above methods e.g.

```python

mask = s.get_signal_mask() # This returns a mask masking nothing by default
mask.set_signal_range() # It remembers its parent mask and, if no signal range is passed, it plots s for interactive selection
mask.set_signal_range(signal=s2) # Interactively edit the mask with a different signal
s.spikes_removal_tool(signal_mask=mask)

```

francisco-dlp commented on hyperspy/_signals/eels.py:
```diff
@@ -299,6 +299,31 @@ def estimate_zero_loss_peak_centre(s, mask, signal_range):
+                            signal_end=None,
+                            filter_zero_loss=False,
+                            zero_loss_width=5.0):
+        if filter_zero_loss:
```
This option is really useful! Could implemented without the ``signal_start`` and ``signal_end`` kwargs?

ericpre replied:
I will make close this one and make a new PR.

francisco-dlp commented on hyperspy/_signals/eels.py:
```diff
@@ -299,6 +299,31 @@ def estimate_zero_loss_peak_centre(s, mask, signal_range):
+                signal_mask = mask & signal_mask
+            else:
+                signal_mask = mask
+            import matplotlib.pyplot as plt
```
We discourage plotting things outside of the signal plot as much as possible. An alternative would be to plot in a different way (maybe using a marker) (or simply not to plot as in model) the masked part.

ericpre replied:
actually, it is some left over of troubleshooting... but it would be good to find a way to display where is the "masked" area.

ericpre committed:
Add filter_zero_loss_peak option to spikes_removal_tool

ericpre committed:
improve spikes_removal_tools documentation

ericpre committed:
- Fix typo in dev_guide
- Add link to the summary of the implementation of lazy signal to the dev_guide

ericpre committed:
add some tests for spikes_removal_tool

ericpre committed:
- Add threshold argument to spikes_removal_tool
- Display signal masked area with fill_between 0 and signal.

ericpre committed:
- update pointer to spikes position
- close plot at the same time than closing the window

ericpre commented:
Ready for review.

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into add_zero_loss_peak_argument_spikes_removal_tool

# Conflicts:
#	doc/dev_guide.rst
#	hyperspy/_signals/signal1d.py
#	hyperspy/gui/egerton_quantification.py

ericpre committed:
merge conflicts: move changes from deleted egerton_quantification.py to signal_tools.py and remove unnecessary tests.

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_patch' into add_zero_loss_peak_argument_spikes_removal_tool

ericpre committed:
Re-introduce tests for spikes_removal_tool

ericpre committed:
- Split `SpikesRemoval` into `SpikesRemoval` and `SpikesRemovalInteractive` to make it interactive or not.
- Allow the threshold to be estimated automatically using the `spikes_diagnosis` tool.
- Update documnentation.

ericpre committed:
Update tests.

ericpre committed:
Fix bug if no spikes is found using the spikes diagnosis tool.

ericpre committed:
Fix issue with passing argument to determine the threshold value to remove spikes.

ericpre committed:
Add more tests to increase coverage and fix one bug with mask for the spikes_removal_tool.

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_patch' into add_zero_loss_peak_argument_spikes_removal_tool

# Conflicts:
#	hyperspy/Release.py
#	hyperspy/tests/signal/test_1D_tools.py

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_patch' into add_zero_loss_peak_argument_spikes_removal_tool

# Conflicts:
#	hyperspy/Release.py
#	hyperspy/_signals/signal1d.py
#	hyperspy/signal_tools.py
#	hyperspy/tests/signal/test_tools.py

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into add_zero_loss_peak_argument_spikes_removal_tool

# Conflicts:
#	doc/dev_guide.rst
#	hyperspy/signal_tools.py

ericpre commented:
Merged with upstream and resolved conflict.

tjof2 commented on doc/user_guide/signal1d.rst:
```diff
@@ -121,8 +121,25 @@ Spike removal
-interface to remove spikes from spectra.
+interface to remove spikes from spectra. It is possible to use this tool
+on a specific interval of the data by :ref:`slicing the data
+<signal.indexing>`. For example, to keep the signal between indexes 8 and 17:
```
I think "keep" is wrong, perhaps "to work only on the signal between indexes 8 and 17"

tjof2 reviewed (commented):
Minor doc change @ericpre otherwise looks good!

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into add_zero_loss_peak_argument_spikes_removal_tool

# Conflicts:
#	doc/dev_guide/lazy_computations.rst
#	doc/user_guide/signal1d.rst

ericpre committed:
Tweak user guide, thanks @tjof2.

tjof2 commented:
If you can resolve the conflicts this should be good to go?

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into add_zero_loss_peak_argument_spikes_removal_tool

# Conflicts:
#	hyperspy/tests/signal/test_tools.py

tjof2 commented:
@francisco-dlp can you check this over again since you initially reviewed it?

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into add_zero_loss_peak_argument_spikes_removal_tool

# Conflicts:
#	hyperspy/signal_tools.py

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into add_zero_loss_peak_argument_spikes_removal_tool

# Conflicts:
#	hyperspy/signal_tools.py

ericpre committed:
Improve and re-use docstrings and tweak name argument.

ericpre committed:
Check zero loss in energy range and fix test.

ericpre committed:
Improve coverage.

ericpre committed:
Fix test when no GUI is installed.

tjof2 reviewed (approved):
Subject to resolving conflicts.

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into add_zero_loss_peak_argument_spikes_removal_tool

# Conflicts:
#	doc/user_guide/signal1d.rst
#	hyperspy/_signals/eels.py
#	hyperspy/_signals/signal1d.py
#	hyperspy/tests/signal/test_eels.py
#	hyperspy/tests/signal/test_tools.py
